x-pack/plugins/security/common/is_internal_url.ts

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { parse, URL } from 'url';
+
+export function isInternalURL(url: string, basePath = '') {
+  const { protocol, hostname, port, pathname } = parse(
+    url,
+    false /* parseQueryString */,
+    true /* slashesDenoteHost */
+  );
+
+  // We should explicitly compare `protocol`, `port` and `hostname` to null to make sure these are not
+  // detected in the URL at all. For example `hostname` can be empty string for Node URL parser, but
+  // browser (because of various bwc reasons) processes URL differently (e.g. `///abc.com` - for browser
+  // hostname is `abc.com`, but for Node hostname is an empty string i.e. everything between schema (`//`)
+  // and the first slash that belongs to path.
+  if (protocol !== null || hostname !== null || port !== null) {
+    return false;
+  }
+
+  if (basePath) {
+    // Now we need to normalize URL to make sure any relative path segments (`..`) cannot escape expected
+    // base path. We can rely on `URL` with a localhost to automatically "normalize" the URL.
+    const normalizedPathname = new URL(String(pathname), 'https://localhost').pathname;
```

Comment:
__note:__ at some point we may want to migrate from legacy `url.parse` API to this WHATWG URL API completely, but I couldn't find a reasonable way to figure out whether URL is relative or absolute just using this new API. The only option I can think of is something like this:
```js
const normalizedURL = new URL(url, 'https://localhost');
const isAbsolute =
  // `true` if `url` has it's own origin different from our "marker" origin
  normalizedURL.origin !== 'https://localhost' ||
  // `true` if `url` has its own origin and it's equal to our "marker" origin
  url.trim().toLowerCase().startsWith('https://localhost');
```

Reply:
I had all of those same thoughts when reviewing this, and decided that it should be out of scope here 😄

Reply:
I'm not able to get this to work locally. When trying to login with a base path, I'm presented with an error:

> url__WEBPACK_IMPORTED_MODULE_0__.URL is not a constructor

![image](https://user-images.githubusercontent.com/3493255/85403165-e6950080-b52a-11ea-9b47-8ab0ef22e0b4.png)

Reply:
Ha, nice find! That's because `URL` got into `import` somehow. It's should be a global instead. But now I'm worried it may not be supported by IE11 - checking..

Reply:
Okay, I've pushed the fix (moved `URL` out of `import` as it should be). I see we use polyfill from CoreJS for the URL so it should work in IE 11. I'll manually pick that fix to 7.8 to check, cause master doesn't work in IE anymore :cry:

Reply:
Just checked on 7.8 - the URL polyfill works fine in IE 11 :phew: